src/main/java/org/candlepin/controller/util/ContentPathBuilder.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.controller.util;
+
+import org.candlepin.model.Environment;
+import org.candlepin.model.Owner;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Class encapsulates the building of full content paths.
+ */
+public class ContentPathBuilder {
+
+    /**
+     * A factory method constructs an instance and prepares prefixes for the given environments.
+     *
+     * @param owner an owner to be used in construction of content prefixes
+     * @param environments environments for which to prepare content prefixes
+     * @return content prefix instance
+     */
+    public static ContentPathBuilder from(Owner owner, List<Environment> environments) {
+        ContentPathBuilder prefixes = new ContentPathBuilder(owner);
+        for (Environment environment : environments) {
+            prefixes.add(environment);
+        }
+        return prefixes;
+    }
+
+    private final Map<String, String> envPrefixes = new HashMap<>();
+    private final Map<String, String> envNames = new HashMap<>();
+    private final String ownerPrefix;
+
+    /**
+     * A constructor
+     *
+     * @param owner an owner to be used in construction of content prefixes
+     */
+    private ContentPathBuilder(Owner owner) {
+        String prefix = owner == null ? "" : owner.getContentPrefix();
+        this.ownerPrefix = normalizePath(prefix);
+    }
+
+    /**
+     * Builds a full content path for the given content and environment.
+     * <p>
+     * Path is constructed in order of content -> environment -> owner. If the
+     * path is already a full path (it includes protocol), the rest of the
+     * prefixes are ignored. E.G. After including environment prefix
+     * 'https://console.redhat.com/template', owner prefix is skipped.
+     * <p>
+     * Builder also populates environment placeholders. Currently only
+     * environment name is supported.
+     *
+     * @param environmentId an id of environment for which to return a content prefix
+     * @param contentPath a content part of the full content path for which to construct the full path.
+     * @return full content path
+     */
+    public String build(String environmentId, String contentPath) {
+        if (contentPath == null || contentPath.isBlank()) {
+            throw new IllegalArgumentException("Content path cannot be null!");
+        }
+
+        String fullPath = buildFullPath(environmentId, contentPath);
+        String populatedPath = populatePlaceholders(fullPath, environmentId);
+        return populatedPath;
+    }
+
+    public String buildFullPath(String environmentId, String contentPath) {
+        String normalizedContentPath = normalizePath(contentPath);
```

Comment:
Normalization here handles only starting/ending slashes. It doesn't touch anything else, although I was wondering whether it would be worth it to handle inner slashes as well. I am not sure how much we can trust the validity of these paths.

Reply:
The current code does the split on forward slashes, and then URL encodes all the chunks, preserving only $ for the various path macros used by DNF. That's the bit we should probably still keep; though we won't want to do that on the scheme or domain if present.

Reply:
Done.